common/services/move.js

```diff
@@ -278,9 +280,17 @@ class MoveService extends BaseService {
       return Promise.reject(new Error(noMoveIdMessage))
     }
 
-    return this.apiClient
-      .find('move', id, options)
-      .then(response => response.data)
+    const { canAccess } = this.req
+
+    return this.apiClient.find('move', id, options).then(response => {
+      const move = response.data
+
+      return {
+        ...move,
+        _canCancel: canCancelMove(move, canAccess),
+        _canEdit: canEditMove(move, canAccess),
```

Comment:
I wasn't completely sure on this naming convention so would be open to any other better suggestions here. 

I've continued to use an underscore prefix to signify that these properties are created within the app as opposed to coming from the API layer. It [something we also do](https://github.com/ministryofjustice/hmpps-book-secure-move-frontend/blob/1c590276f91c00cfd7443f821d1484527f17b8bc/common/lib/api-client/transformers/person.transformer.js#L2) in the resource transformers for the JSON:API serialization layer.

I toyed with a few different names in my head too so would be good to get opinions on what maybe works better: `_canEdit`, `_isEditable`, `_editable` ...?

Reply:
I'm happy with `canEdit` since it matches the prefix of the `canEditMove` function and implies they're related (which they are!).

I'm unsure about the underscores myself and even whether this is the right place to insert the `canEdit` information. Without knowing more than the function name, I'm not sure you'd expect the data from the API to be transformed to include this information. I suppose this is why we've got the `_` convention, but there's nothing that actually stops an API response from including fields which start with an underscore. I was also a little surprised to see `this.req` being available in what to me seems like a class that shouldn't necessarily know about the current request, but I think this would be a wider change of this service!

Without knowing the architecture of the app too well, is there a placer higher up where we can insert this field?

Since we've already got this convention in the app, I'm also happy to accept this as it is, better to stick with one way of doing things if we're already doing it in lots of places.

Reply:
>I'm happy with canEdit since it matches the prefix of the canEditMove function and implies they're related (which they are!).

👍 

>Without knowing more than the function name, I'm not sure you'd expect the data from the API to be transformed to include this information. I suppose this is why we've got the _ convention, but there's nothing that actually stops an API response from including fields which start with an underscore.

Yes, that's correct, there is nothing stopping the API including an underscore but I suppose its also just a convention the API has stayed away from which is why we adopted it just for this layer. It allowed us to insert extra data that the API didn't need to or couldn't necessarily insert for us.

And that's actually the case with this example because the editability within this frontend layer is determined by a user's permissions as well as the particular state of the move, and with the architecture we currently have the API doesn't know about the current user's permissions. So it has to be done somewhere in this layer. 

>I'm unsure about the underscores myself and even whether this is the right place to insert the canEdit information.

This is the part that follows on from the above, previously we were doing it in the `getUpdateUrls` helper but that required the `canAccess` method to be passed down through various different layers and functions. This is what I was trying to remove to make some of those functions a little simpler.

But the main problem is that to determine whether it can be edited by this particular user we need access to the user's permissions which lives on the session and in turn the `req` object. So we can either do it in lower level middleware or controllers (this was the previous implementation) or somewhere higher like this move service.

>I was also a little surprised to see `this.req` being available in what to me seems like a class that shouldn't necessarily know about the current request, but I think this would be a wider change of this service!

There is a bit of backstory to this change. The services didn't previously have access to the request object but it was a change done to allow all services have access to it so that we could pass relevant information from the request [along with each API request](https://github.com/ministryofjustice/hmpps-book-secure-move-frontend/blob/main/common/lib/api-client/request-headers.js), for example a user's name, ID for audit or transaction IDs to allow us to link up frontend and API requests or errors. 

Here's [a change](https://github.com/ministryofjustice/hmpps-book-secure-move-frontend/pull/1124/) I found related to that but it was done over a series of PRs. In hindsight, it would have been nice to use a particular issue/milestone/project so that we could've looked back at all related changes easier.

With a bit more context how does the current placement within this service sound?

Reply:
Thanks for the detailed response, that makes a lot of sense. Happy to proceed with the new placement, it's definitely an improvement on how it is now!